File: Python/count-beautiful-substrings-ii.py
# Time:  O(n + sqrt(k))
# Space: O(1)

import collections


# number theory, prefix sum, freq table
class Solution(object):
    def beautifulSubstrings(self, s, k):
        """
        :type s: str
        :type k: int
        :rtype: int
        """
        VOWELS = set("aeiou")
        prefix = [0]*(len(s)+1)
        for i in xrange(len(s)):
            prefix[i+1] = prefix[i]+(+1 if s[i] in VOWELS else -1)
        new_k = 1
        x = k
        for i in xrange(2, k+1):
            if i*i > k:
                break
            cnt = 0
            while x%i == 0:
                x //= i
                cnt += 1
            if cnt:
                new_k *= i**((cnt+1)//2+int(i == 2))
        if x != 1:
            new_k *= x**((1+1)//2+int(x == 2))
        cnt = collections.Counter()
        result = 0
        for i, p in enumerate(prefix):
            result += cnt[p, i%new_k]
            cnt[p, i%new_k] += 1
        return result

File: C++/count-beautiful-substrings-ii.cpp
// Time:  O(n + sqrt(k))
// Space: O(n)

// number theory, prefix sum, freq table
class Solution {
public:
    int beautifulSubstrings(string s, int k) {
        static const unordered_set<char> VOWELS = {'a', 'e', 'i', 'o', 'u'};
    
        vector<int> prefix(size(s) + 1);
        for (int i = 0; i < size(s); ++i) {
            prefix[i + 1] = prefix[i] + (VOWELS.count(s[i]) ? +1 : -1);
        }
        int new_k = 1;
        int x = k;
        for (int i = 2; i * i <= k; ++i) {
            int cnt = 0;
            for (; x % i == 0; x /= i) {
                ++cnt;
            }
            if (cnt) {
                new_k *= pow(i, (cnt + 1) / 2 + static_cast<int>(i == 2));
            }
        }
        if (x != 1) {
            new_k *= pow(x, (1 + 1) / 2 + static_cast<int>(x == 2));
        }
        unordered_map<int, unordered_map<int, int>> cnt;
        int result = 0;
        for (int i = 0; i < size(prefix); ++i) {
            result += cnt[prefix[i]][i % new_k]++;
        }
        return result;
    }
};

